Developer docs: add multi-component development example (#68)
ACCESS-NRI/ACCESS-OM2: DEVELOPERS.md

git clone -c feature.manyFiles=true https://github.com/ACCESS-NRI/spack.git --branch releases/v0.21 --single-branch --depth=1
git clone https://github.com/ACCESS-NRI/spack-packages.git --branch main
git clone https://github.com/ACCESS-NRI/spack-config.git --branch main
```

Then link all the `spack-config` settings to your local `spack` instance:
```bash
ln -s -r -v spack-config/v0.21/gadi/* spack/etc/spack/
```

> [!NOTE]
>
> The steps in this section need only be done once, unless the cloned repositories need to be updated to a newer version.
>
> This guide is adapted from [instructions on the ACCESS-Hive Forum](https://forum.access-hive.org.au/t/how-to-build-access-om2-on-gadi/1545)
>
> `spack` is cloned from the ACCESS-NRI fork as on occasion there are some fixes that are back-ported by ACCESS-NRI from more recent versions of `spack`.

## Enable spack

For every new login or new shell it is necessary to add your local copy of `spack` to your shell, as well as some other settings. 
```bash
. spack-config/spack-enable.bash
```
> [!NOTE]
> There is a space between the `.` and the path, as we are not treating `spack-enable.bash` as an executable

## Development environment

> Spack [has environments](https://spack.readthedocs.io/en/latest/environments.html), which is similar to many other related implementations such as conda environments and python virtual environments.

Reviewer: ```suggestion
Spack [has environments](https://spack.readthedocs.io/en/latest/environments.html), which are similar to many other related implementations such as conda environments and python virtual environments.
```

Choose a name that suits the aim of the work. Make a directory with that name, `cd` into that directory then type the command `spack env create -d .`. This creates a `spack` independent environment in the current directory.

### Modifying an environment

Once the environment is created type `spack env activate .` from within the environment directory to activate the environment. 

Add packages that are required, and use `spack develop` to indicate which packages will be modified and recompiled. 

By default `spack` will clone the source code automatically, but it is possible to use an existing source code repository directory if that is preferred.

It is necessary after `spack develop` to call `spack concretize -f` to force spack to update the concretization so that it picks up the changes to the packages that are being developed.

## Compiling Modified Code

Now the source code can be modified, and then compiled by invoking `spack install`

## Examples

The instructions above are best understood by example.

### Developing a single model component

In this example the `mom5` component will be modified but all other components and dependencies remain the same. 

> This example also uses the [`ACCESS-OM2` `spack.yaml` environment file](https://github.com/ACCESS-NRI/ACCESS-OM2/blob/main/spack.yaml) as an argument to the [`spack env create` command](https://spack.readthedocs.io/en/latest/command_index.html#spack-env-create). This makes the development environment the same as that for the released ACCESS-OM2 model that is referenced in the local copy of the repo. So it is necessary to clone [this repository](https://github.com/ACCESS-NRI/ACCESS-OM2/).

Reviewer: ```suggestion
This example also uses the [`ACCESS-OM2` `spack.yaml` environment file](https://github.com/ACCESS-NRI/ACCESS-OM2/blob/main/spack.yaml) as an argument to the [`spack env create` command](https://spack.readthedocs.io/en/latest/command_index.html#spack-env-create). This makes the development environment the same as that for the released ACCESS-OM2 model that is referenced in the local copy of the repo. So it is necessary to clone [the ACCESS-NRI ACCESS-OM2 repository](https://github.com/ACCESS-NRI/ACCESS-OM2/).
```

[+] /g/data/vk83/apps/spack/0.20/release/linux-rocky8-x86_64/intel-19.0.5.281/cice5-git.2023.10.19=2023.10.19-v3zncpqjj2gyseudbwiudolcjq3k3leo
==> Installing access-om2-git.2024.03.0=2024.03.0-5nk4sjpusp6cilzp3jkmubonty5q2w62
==> No binary for access-om2-git.2024.03.0=2024.03.0-5nk4sjpusp6cilzp3jkmubonty5q2w62 found: installing from source
==> No patches needed for access-om2
==> access-om2: Executing phase: 'install'
==> access-om2: Successfully installed access-om2-git.2024.03.0=2024.03.0-5nk4sjpusp6cilzp3jkmubonty5q2w62
  Stage: 0.01s.  Install: 0.00s.  Post-install: 0.12s.  Total: 0.86s
[+] /g/data/.../dev_instructions/release/linux-rocky8-x86_64/intel-19.0.5.281/access-om2-git.2024.03.0=2024.03.0-5nk4sjpusp6cilzp3jkmubonty5q2w62
==> Updating view at /g/data/.../dev_instructions/envs/mom5-dev/.spack-env/view
==> Warning: Skipping external package: openmpi@4.0.2%intel@19.0.5.281~atomics~cuda~cxx~cxx_exceptions~gpfs~internal-hwloc~java~legacylaunchers~lustre~memchecker~orterunprefix+romio+rsh~singularity+static+vt+wrapper-rpath build_system=autotools fabrics=none patches=073477a,60ce20b schedulers=none arch=linux-rocky8-x86_64/4jwtg3n
```

> 

Reviewer: say something here about where the newly-compiled executables can be found

==> Warning: Module file /g/data/.../dev_instructions/release/modules/linux-rocky8-x86_64/access-om2/2024.03.0 exists and will not be overwritten
==> Warning: Module file /g/data/.../dev_instructions/release/modules/linux-rocky8-x86_64/mom5/2023.11.09 exists and will not be overwritten
==> Error: Installation request failed.  Refer to reported errors for failing package(s).
```

6. Diagnose build error and rebuild

As expected the change resulted in an error and the build was not successful.

`spack` lets the user know where the build output and errors are viewable (`mom5/spack-build-out.txt`), but in this case the error is obvious:

```bash
  >> 161    /g/data/.../dev_instructions/envs/mom5-dev/mom5/src/accessom_coupler/ocean_solo.F90(365): error #6404: This name does not have a type, and must have an explicit type.   [SECONDS_PER_HOUR]
     162        sfix_seconds = sfix_hours * SECONDS_PER_HOUR
     163    --------------------------------^
     164    compilation aborted for /g/data/.../dev_instructions/envs/mom5-dev/mom5/src/accessom_coupler/ocean_solo.F90 (code 1)
```

As a result of the intentional error introduced above. 

> Reverting that change and `spack install` worked as expected

Reviewer: ```suggestion
Reverting that change, `spack install` works as expected
```
